src/amici.cpp

```diff
@@ -2185,3 +2192,90 @@ void processUserData(UserData *udata) {
     dwdp_tmp = new realtype[ndwdp]();
 }
+
+int fsy(realtype t_, int it, realtype *sy, realtype *dydx_, realtype *dydp_, N_Vector *sx, void *user_data){
+    // Compute sy = dydx * sx + dydp
+
+    int status = 0;
+    UserData *udata = (UserData*) user_data;
+
+    for(int ip = 0; ip < nplist; ++ip) {
+        for(int iy = 0; iy < ny; ++iy)
+            // copy dydp to sy
+            sy[ip * nt * ny + iy * nt + it] = dydp_[iy + ip * ny];
+
+        realtype *sxTmp = N_VGetArrayPointer(sx[ip]);
+
+        // compute sy = 1.0*dydx*sx + 1.0*sy
+        cblas_dgemv(CblasColMajor, CblasNoTrans, ny, nx,
+                    1.0, dydx_, ny, sxTmp, 1,
+                    1.0, &sy[ip * nt * ny + it], nt);
+    }
+
+    return status;
+}
+
+int fsJy(realtype t_, int it, realtype *sJy, realtype *s2Jy, realtype *dJydy, realtype *dJydp, realtype *y, realtype *sigma_y_, realtype *sy, realtype *dydp_, realtype *my_, void *user_data) {
+    int status = 0;
+    UserData *udata = (UserData*) user_data;
+
+    // Compute sy-dydp for current 'it'
+    // dydp         ny x nplist
+    // sy           nt x ny x nṕlist
+    // dydp part needs to be substracted as it is already contained in dJydp
+    // we only need to account for sensitivities here
+    realtype *diff = new realtype[ny * nplist];
+    for(int iy = 0; iy < ny; ++iy)
+        for(int ip = 0; ip < nplist; ++ip)
+            diff[iy + ip * ny] = sy[ip * nt * ny + iy * nt + it] - dydp_[iy + ip * ny];
+
+    // sJy          nplist x ng
+    // dJydp=dgdp   nytrue x nplist x ng
+    // dJydy=dgdy   nytrue x ng x ny
+
+    realtype *dJydyTmp = new realtype[ng * ny];
+    realtype *multResult = new realtype[nplist * ng];
+
+    for(int iyt = 0; iyt < nytrue; ++iyt) {
+        if(amiIsNaN(my_[nt * iyt + it]))
+            continue;
+
+        // copy current (iyt) dJydy slice
+        // dJydyTmp     ng x ny
+        for(int ig = 0; ig < ng; ++ig)
+            for(int iy = 0; iy < ny; ++iy)
+                dJydyTmp[ig + iy * ng] = dJydy[iyt + ig * nytrue + iy * nytrue * ng];
+
+        // compute multResult = (dJydyTmp * diff)' + dJydp == diff' * dJydyTmp' + dJydp
+        // copy dJydp slice (iyt) to result
+        for(int ip = 0; ip < nplist; ++ip)
+            for(int ig = 0; ig < ng; ++ig)
+                multResult[ip + np * ig] = dJydp[iyt + ip * nytrue + ig * nytrue * nplist];
+
+        // C := alpha*op(A)*op(B) + beta*C,
+        cblas_dgemm(CblasColMajor, CblasTrans, CblasTrans,
+                    nplist, ng, ny,
+                    1.0, diff, ny,
+                    dJydyTmp, ng,
+                    1.0, multResult, nplist);
+
+
+        // sJy += multResult
+        for(int ig = 0; ig < ng; ++ig) {
+            if(ig == 0)
+                for(int ip = 0; ip < nplist; ++ip)
+                    sJy[ip] -= multResult[ip];
+            else
+                for(int ip = 0; ip < nplist; ++ip)
+                    s2Jy[ip + nplist * ig] -= multResult[ip+ nplist * ig];
```

Comment:
Right. Well spotted, even before the test failed. Which did fail, luckily. Done: 464734e.